app/controllers/users/sessions_controller.rb

```diff
@@ -4,12 +4,25 @@ class SessionsController < Devise::SessionsController
 
     skip_before_action :session_expires_at, only: [:active]
     skip_after_action :track_get_requests, only: [:active]
+    before_action :confirm_two_factor_authenticated, only: [:update]
 
     def create
       track_authentication_attempt(params[:user][:email])
       super
     end
 
+    def show
+    end
+
+    def update
+      if current_user.valid_password?(password)
+        redirect_to user_two_factor_authentication_path(reauthn: true)
+      else
+        flash[:error] = t('errors.confirm_password_incorrect')
+        redirect_to user_password_confirm_path
+      end
+    end
```

Comment:
It's not the password that needs validating. It's the whole 2FA mechanism. That's why I put it in the session controller: we are updating the session to flag it as "recently authenticated".